### PR TITLE
change keybased channel to multi-keys version

### PR DESCRIPTION
**Pull request by rogercloud**

### CODE REVIEW

File: curp/src/keybased_channel.rs [hunk 1]
```diff
@@ -12,55 +13,90 @@ use tokio::task::JoinHandle;
...
+    /// Create a new `KeysMessageInner` structure
+    fn new(keys: Vec<K>, message: M) -> Self {
+        Self { keys, message }
+    }
+}
+
+/// Keys and Messages combined structure
+pub(crate) struct KeysMessage<K, M> {
+    /// Inner
+    inner: Arc<KeysMessageInner<K, M>>,
```
Comment (pzheng1025) [inline curp/src/keybased_channel.rs:34]: Is it possible to use Rc instead of Arc?

Comment (rogercloud) [reply curp/src/keybased_channel.rs:34]: It's transfered via async channels, Arc is required.

File: curp/src/keybased_channel.rs [hunk 2]
```diff
@@ -118,31 +220,26 @@ enum MessageOrListener<M> {
...
-                Some(msg) => return MessageOrListener::Message(msg),
-                #[allow(clippy::match_ref_pats)] // don't move listener out
-                None => match &local_listener {
-                    &Some(_) => {
-                        break;
-                    }
-                    &None => local_listener = Some(channel.new_msg_event.listen()),
-                },
-            }
+        self.buf.extend(channel.pop_all_msg().into_iter());
```
Comment (pzheng1025) [inline curp/src/keybased_channel.rs:232]: possible to use mem::swap()?

Comment (rogercloud) [reply curp/src/keybased_channel.rs:232]: it's just one move fromone place to another, mem::swap is not more efficient in this case I think.

**Review by pzheng1025 — APPROVED** [review]

**Pull request merged**